lib/hal/portable/st/aws_hal_perf_cntr.c

```diff
@@ -0,0 +1,233 @@
+{
+    HAL_TIM_IRQHandler( &xTimerHandle );
+}
+/*-----------------------------------------------------------*/
+
+void aws_hal_perfcounter_open(void)
+{
+    uint32_t ulTimClock = 0;
+
+    /* Clear static variable which tracks total number of timer overflow. */
+    ullTimerOverflow = 0;
+
+    /* Get timer clock frequency. */
+    ulTimClock = HAL_RCC_GetPCLK1Freq();
+
+    /* Check user input frequency is valid. */
+    configASSERT( HAL_PERF_COUNTER_FREQ <= ulTimClock );
+
+    /* From STM32L475xx chip datasheet:
+     * - TIM2 and TIM5 are 32-bit general purpose timers, while others are all 16-bit.
+     * - Peripheral current consumption (Range 1/Range 2/Low power and sleep):
+     *   TIM2: 6.8/6.5/6.3 (uA/MHZ)
+     *   TIM5: 6.5/5.5/6.1 (uA/MHZ)
+     * To minimize the need for handling timer overflow interrupt with relative low cost,
+     * prefer to use TIM5 as the hardware timer for performance counter.
+     */
+    xTimerHandle.Instance = TIM5;
+
+    /* To minimize the need for serving overflow IRQ,
+     * .Period is always set to the max value the counter allows.
+     * Performance counter resolution is adjusted with .Prescaler.
+     */
+    xTimerHandle.Init.Prescaler = ulTimClock / HAL_PERF_COUNTER_FREQ;
+    xTimerHandle.Init.CounterMode = TIM_COUNTERMODE_UP;
+    xTimerHandle.Init.Period = HW_TIMER_32_CONST_PERIOD;
+    xTimerHandle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+    xTimerHandle.Init.RepetitionCounter = 0x00;
+    xTimerHandle.Init.AutoReloadPreload = TIM_AUTORELOAD_PRELOAD_ENABLE;
+
+    xTimerHandle.Channel = HAL_TIM_ACTIVE_CHANNEL_CLEARED;
+
+    xTimerHandle.Lock = HAL_UNLOCKED;
+
+    xTimerHandle.State = HAL_TIM_STATE_RESET;
+
+    HAL_TIM_Base_Init( &xTimerHandle );
+
+    /* Only generate interrupt when overflow. */
+    __HAL_TIM_URS_ENABLE( &xTimerHandle );
+
+    /* Clear the initial SR.UIF bit before timer starts. */
+    __HAL_TIM_CLEAR_FLAG( &xTimerHandle, TIM_FLAG_UPDATE);
+
+    /* Register overflow IRQ callback. */
+    HAL_TIM_RegisterCallback( &xTimerHandle, HAL_TIM_PERIOD_ELAPSED_CB_ID, HAL_TIM5_PeriodElapsedCallback );
+
+    /* Timer and interrupt are configured. Now start timer. */
+    HAL_TIM_Base_Start_IT( &xTimerHandle );
+}
+
+/*-----------------------------------------------------------*/
+
+void aws_hal_perfcounter_close( void)
+{
+    HAL_TIM_Base_Stop_IT( &xTimerHandle );
+
+    HAL_TIM_Base_DeInit( &xTimerHandle );
+}
+
+/*-----------------------------------------------------------*/
+
+uint64_t aws_hal_perfcounter_get_value(void)
+{
+    return ( uint64_t ) ( ( ullTimerOverflow << HW_TIMER_32_WIDTH ) | __HAL_TIM_GET_COUNTER( &xTimerHandle ) );
+}
+
+/*-----------------------------------------------------------*/
+
+uint32_t aws_hal_perfcounter_get_frequency_hz(void)
+{
+    uint32_t ulTimClock = 0;
+
+    /* Get timer clock frequency. */
+    ulTimClock = HAL_RCC_GetPCLK1Freq();
+
+    return ( ulTimClock / ( uint32_t )( ulTimClock / HAL_PERF_COUNTER_FREQ ) );
```

Comment:
ulTimClock in STM32L4xx's context is the frequency of PCLK1 which comes from APB1 (Advanced Peripheral Bus). In order to make ```HAL_RCC_GetPCLK1Freq()``` return zero, from ST clock tree diagram it requires the entire bus to be shut down, which impacts TIM/RTC/UART/I2C/...... Correct me here, I think the sane way is to keep this bus clock on always, and turn on/off individual components on demand, which makes ```HAL_RCC_GetPCLK1Freq()``` not zero... 

As for "Isn't it just HAL_PERF_COUNTER_FREQ?" It may be easier for people to read if whatever user input for ```HAL_PERF_COUNTER_FREQ ``` in FreeRTOSConfig.h actually reflects the counter frequency. Right not ```HAL_PERF_COUNTER_FREQ``` does take whatever value user inputs, but .prescaler is always set to the closest integer value. (line 177 in the same file.) The trade off made here is that timer doesn't always get the resolution user requested. (e.g. Would 12.5ns vs. 14ns per clock cycle really makes a difference? Probably not... ) But calculation is simpler and the need to serve IRQ is minimized with .period always set to maximum value.